src/SourceBuild/content/eng/tools/tasks/Microsoft.DotNet.UnifiedBuild.Tasks/AzureDevOpsClient.cs

```diff
@@ -0,0 +1,193 @@
+                "Basic",
+                Convert.ToBase64String(Encoding.UTF8.GetBytes($":{azureDevOpsToken}")));
+        }
+    }
+
+    /// <summary>
+    /// Downloads specified packages and symbols from a specific build artifact and stores them in an output folder, either flat or with the same relative path as in the artifact.
+    /// </summary>
+    public async Task DownloadArtifactFiles(
+        string buildId,
+        string artifactName,
+        List<string> packageNames,
+        List<string> symbolNames,
+        string downloadFolder,
+        string outputFolder,
+        bool flatCopy)
+    {
+        string downloadPath = Path.Combine(downloadFolder, "artifact.zip");
+        string extractPath = Path.Combine(downloadFolder, "extracted");
+        string extractedAssetsPath = Path.Combine(extractPath, artifactName, _assetsFolderName);
+        string extractedPackagesPath = Path.Combine(extractPath, artifactName, _packagesFolderName);
+
+        try{
+            await DownloadArtifactZip(buildId, artifactName, downloadPath);
+
+            ZipFile.ExtractToDirectory(downloadPath, extractPath);
+
+            // assets include all kinds of files, so just look at all files
+            List<string> sourceSymbolPaths = Directory.GetFiles(extractedAssetsPath, "*", SearchOption.AllDirectories).ToList();
+            List<string> sourcePackagePaths = Directory.GetFiles(extractedPackagesPath, "*.nupkg", SearchOption.AllDirectories).ToList();
+        
+            string packageOutputPath = Path.Combine(outputFolder, _packagesFolderName);
+            string symbolOutputPath = Path.Combine(outputFolder, _assetsFolderName);
+
+            if (!Directory.Exists(packageOutputPath))
+            {
+                Directory.CreateDirectory(packageOutputPath);
+            }
+            if (!Directory.Exists(symbolOutputPath))
+            {
+                Directory.CreateDirectory(symbolOutputPath);
+            }
+
+            CopyFiles(packageNames, sourcePackagePaths, extractedPackagesPath, packageOutputPath, flatCopy);
+            CopyFiles(symbolNames, sourceSymbolPaths, extractedAssetsPath, symbolOutputPath, flatCopy);
+        }
+        finally
+        {
+            if (Directory.Exists(extractPath))
+            {
+                Directory.Delete(extractPath, true);
+            }
+            if (File.Exists(downloadPath))
+            {
+                File.Delete(downloadPath);
+            }
+        }
+    }
+
+    private async Task DownloadArtifactZip(string buildId, string artifactName, string downloadPath) 
+    {
+        string relativeUrl = $"build/builds/{buildId}/artifacts?artifactName={artifactName}&api-version={_azureDevOpsApiVersion}";
+        _logger.LogMessage(MessageImportance.High, $"Downloading artifact information from {relativeUrl}");
+        HttpResponseMessage response = await _httpClient.GetAsync(relativeUrl);
```

Comment:
This functionality probably needs a try/catch w/retry on certain responses.

Reply:
A common pattern is `EnsureSuccessStatusCode()` here with the corresponding try/catch and retry